maas/resource_maas_volume_group.go

```diff
@@ -0,0 +1,259 @@
+	}
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	volumeGroup, err := client.VolumeGroup.Get(machine.SystemID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	blockDevices := findVolumeGroupBlockDevices(volumeGroup)
+
+	tfState := map[string]interface{}{
+		"block_devices":  blockDevices,
+		"machine":        volumeGroup.SystemID,
+		"name":           volumeGroup.Name,
+		"size":           volumeGroup.Size,
+		"used_size":      volumeGroup.UsedSize,
+		"available_size": volumeGroup.AvailableSize,
+		"uuid":           volumeGroup.UUID,
+	}
+
+	if err := setTerraformState(d, tfState); err != nil {
+		return diag.Errorf("Could not set volume group state: %v", err)
+	}
+
+	return nil
+}
+
+func resourceMaasVolumeGroupUpdate(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	machine, err := getMachine(client, d.Get("machine").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	var addBlockDevices []string
+	var removeBlockDevices []string
+
+	if d.HasChange("block_devices") {
+		oldBlockDevices, newBlockDevices := d.GetChange("block_devices")
+
+		oldDeviceList := convertToStringSlice(oldBlockDevices.(*schema.Set).List())
+		newDeviceList := convertToStringSlice(newBlockDevices.(*schema.Set).List())
+
+		for _, device := range newDeviceList {
+			if !slices.Contains(oldDeviceList, device) {
+				addBlockDevices = append(addBlockDevices, device)
+			}
+		}
+		for _, device := range oldDeviceList {
+			if !slices.Contains(newDeviceList, device) {
+				removeBlockDevices = append(removeBlockDevices, device)
+			}
+		}
+	}
+
+	updateParams := entity.VolumeGroupUpdateParams{
+		Name:               d.Get("name").(string),
+		AddBlockDevices:    addBlockDevices,
+		RemoveBlockDevices: removeBlockDevices,
+	}
+
+	volumeGroup, err := client.VolumeGroup.Update(machine.SystemID, id, &updateParams)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	d.SetId(fmt.Sprintf("%v", volumeGroup.ID))
+
+	return resourceMaasVolumeGroupRead(ctx, d, meta)
+}
+
+func resourceMaasVolumeGroupDelete(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*ClientConfig).Client
+
+	machine, err := getMachine(client, d.Get("machine").(string))
+	if err != nil {
+		return diag.FromErr(err)
+	}
+	id, err := strconv.Atoi(d.Id())
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	err = client.VolumeGroup.Delete(machine.SystemID, id)
+	if err != nil {
+		return diag.FromErr(err)
+	}
+
+	return nil
+}
+
+func findVolumeGroupBlockDevices(volumeGroup *entity.VolumeGroup) []string {
+	var blockDevices []string
+	for _, device := range volumeGroup.Devices.([]interface{}) {
+		thisDevice := device.(map[string]interface{})
+
+		var deviceId string
+
+		// partitions list a device id of the parent block device
```

Comment:
This is not clear to me. So we can create logical groups which contain partitions by providing the id of the partition and in the read response, instead of that id, we get the id of the parent block device? Maybe I miss-understand it. Could you please provide an example with some JSON responses?